# Aviation Risk Analysis

## 📊 Data-Driven Strategy to entering the aviation industry
This project uses the National Transportation Safety Board aviation accident data from 1962 to 2023 in United States and international waters. We will use this to identify strategic opportunities, risks and patterns.

Through exploratory data analysis, trends and visualizations, we will give concrete data-driven business recommendations. This will help investors, insurers assess the risk exposurers and make informed decisions that are low risk.


## ❓ Key Business Questions
As we seek the answers to the questions below, the questions will guide the direction of my project
1. Which aircraft make are least frequently involved in accidents
    - This helps decide on the make or manufacturers(amateur built)
2. What are the most common causes of aviation accidents
    - This helps to understand risk trends, are newer models safer than older one
3. What percentage of accidents are survivable and what factors influence survivability?
    - Here, we will check on engines, weather
4. How do accident rates vary by operator type (commercial vs private)
    - This help to decide the aviation sector to indulge in
5. Are there seasonal patterns in aviation accidents
    - This helps in weather contingency planning
6. What factors contribute to fatal accidents(weather, broad phase of flight)
    - Helps decide on aircraft specs

## 📂Data Understanding

In this section, we examine the datasets used in the project from the National Transportation Safety Board aviation accident data from [Our Dataset](https://www.kaggle.com/datasets/khsamaha/aviation-accident-database-synopses). We familiarize ourselves with the dataset structure. We thoroughly go through the columns and rows, identify missing data, categorical grouping and data range.

We analyze the data quality by checking for completeness,consistency, uniqueness and distributions. A clean dataset gives clear and accurate analysis and visualizations.



In [1]:
# Importing data
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns


In [2]:
#import the csv file AviationData
AviData = pd.read_csv('Data\AviationData.csv', encoding= 'latin1', low_memory= False)

In [3]:
# The first 5 rows
AviData.head()

Event.Id Investigation.Type Accident.Number  Event.Date  \
0  20001218X45444           Accident      SEA87LA080  1948-10-24   
1  20001218X45447           Accident      LAX94LA336  1962-07-19   
2  20061025X01555           Accident      NYC07LA005  1974-08-30   
3  20001218X45448           Accident      LAX96LA321  1977-06-19   
4  20041105X01764           Accident      CHI79FA064  1979-08-02   

          Location        Country   Latitude   Longitude Airport.Code  \
0  MOOSE CREEK, ID  United States        NaN         NaN          NaN   
1   BRIDGEPORT, CA  United States        NaN         NaN          NaN   
2    Saltville, VA  United States  36.922223  -81.878056          NaN   
3       EUREKA, CA  United States        NaN         NaN          NaN   
4       Canton, OH  United States        NaN         NaN          NaN   

  Airport.Name  ... Purpose.of.flight Air.carrier Total.Fatal.Injuries  \
0          NaN  ...          Personal         NaN                  2.0   
1          NaN  ...          Personal         NaN                  4.0   
2          NaN  ...          Personal         NaN                  3.0   
3          NaN  ...          Personal         NaN                  2.0   
4          NaN  ...          Personal         NaN                  1.0   

  Total.Serious.Injuries Total.Minor.Injuries Total.Uninjured  \
0                    0.0                  0.0             0.0   
1                    0.0                  0.0             0.0   
2                    NaN                  NaN             NaN   
3                    0.0                  0.0             0.0   
4                    2.0                  NaN             0.0   

  Weather.Condition  Broad.phase.of.flight   Report.Status Publication.Date  
0               UNK                 Cruise  Probable Cause              NaN  
1               UNK                Unknown  Probable Cause       19-09-1996  
2               IMC                 Cruise  Probable Cause       26-02-2007  
3               IMC                 Cruise  Probable Cause       12-09-2000  
4               VMC               Approach  Probable Cause       16-04-1980  

[5 rows x 31 columns]

In [4]:
# column names#
AviData.columns

Index(['Event.Id', 'Investigation.Type', 'Accident.Number', 'Event.Date',
       'Location', 'Country', 'Latitude', 'Longitude', 'Airport.Code',
       'Airport.Name', 'Injury.Severity', 'Aircraft.damage',
       'Aircraft.Category', 'Registration.Number', 'Make', 'Model',
       'Amateur.Built', 'Number.of.Engines', 'Engine.Type', 'FAR.Description',
       'Schedule', 'Purpose.of.flight', 'Air.carrier', 'Total.Fatal.Injuries',
       'Total.Serious.Injuries', 'Total.Minor.Injuries', 'Total.Uninjured',
       'Weather.Condition', 'Broad.phase.of.flight', 'Report.Status',
       'Publication.Date'],
      dtype='object')

In [5]:
# Information on the dataset
AviData.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 88889 entries, 0 to 88888
Data columns (total 31 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Event.Id                88889 non-null  object 
 1   Investigation.Type      88889 non-null  object 
 2   Accident.Number         88889 non-null  object 
 3   Event.Date              88889 non-null  object 
 4   Location                88837 non-null  object 
 5   Country                 88663 non-null  object 
 6   Latitude                34382 non-null  object 
 7   Longitude               34373 non-null  object 
 8   Airport.Code            50249 non-null  object 
 9   Airport.Name            52790 non-null  object 
 10  Injury.Severity         87889 non-null  object 
 11  Aircraft.damage         85695 non-null  object 
 12  Aircraft.Category       32287 non-null  object 
 13  Registration.Number     87572 non-null  object 
 14  Make                    88826 non-null

In [6]:
# shape of the dataset
AviData.shape

(88889, 31)

In [ ]:
To check missing values in percentage from large to small
AviData.isna().mean().sort_values(ascending=False)

Schedule                  0.858453
Air.carrier               0.812710
FAR.Description           0.639742
Aircraft.Category         0.636772
Longitude                 0.613304
Latitude                  0.613203
Airport.Code              0.434699
Airport.Name              0.406113
Broad.phase.of.flight     0.305606
Publication.Date          0.154924
Total.Serious.Injuries    0.140737
Total.Minor.Injuries      0.134246
Total.Fatal.Injuries      0.128261
Engine.Type               0.079616
Report.Status             0.071786
Purpose.of.flight         0.069660
Number.of.Engines         0.068445
Total.Uninjured           0.066510
Weather.Condition         0.050535
Aircraft.damage           0.035932
Registration.Number       0.014816
Injury.Severity           0.011250
Country                   0.002542
Amateur.Built             0.001147
Model                     0.001035
Make                      0.000709
Location                  0.000585
Event.Date                0.000000
Accident.Number     

## Observation

- We have 31 columns and 88889 rows.
- Most of our columns have categorical data; some of which we will have to change
- Most of our columns have missing data, a few have more than 60% missing data

#### Point to note
We will have to go through each column to decide wether to replace or fill
- We will use fillna() or replace()
- We use 'unknown' for categorical data and mean or 0 for numerical data
- We can also drop columns using dropna()

#### Next step
 - We make a copy before any changes to preserve the original dataset
 - Go through each column and find if it's important in our analysis
 - Identify what to drop
 - 






In [8]:
#copy of the original data before we start dropping 
data = AviData.copy(deep=True)
data.shape

(88889, 31)

In [9]:
data.columns

Index(['Event.Id', 'Investigation.Type', 'Accident.Number', 'Event.Date',
       'Location', 'Country', 'Latitude', 'Longitude', 'Airport.Code',
       'Airport.Name', 'Injury.Severity', 'Aircraft.damage',
       'Aircraft.Category', 'Registration.Number', 'Make', 'Model',
       'Amateur.Built', 'Number.of.Engines', 'Engine.Type', 'FAR.Description',
       'Schedule', 'Purpose.of.flight', 'Air.carrier', 'Total.Fatal.Injuries',
       'Total.Serious.Injuries', 'Total.Minor.Injuries', 'Total.Uninjured',
       'Weather.Condition', 'Broad.phase.of.flight', 'Report.Status',
       'Publication.Date'],
      dtype='object')

In [10]:
# normalize column names
data.columns= data.columns.str.strip().str.lower().str.replace('.','_')
data.columns


Index(['event_id', 'investigation_type', 'accident_number', 'event_date',
       'location', 'country', 'latitude', 'longitude', 'airport_code',
       'airport_name', 'injury_severity', 'aircraft_damage',
       'aircraft_category', 'registration_number', 'make', 'model',
       'amateur_built', 'number_of_engines', 'engine_type', 'far_description',
       'schedule', 'purpose_of_flight', 'air_carrier', 'total_fatal_injuries',
       'total_serious_injuries', 'total_minor_injuries', 'total_uninjured',
       'weather_condition', 'broad_phase_of_flight', 'report_status',
       'publication_date'],
      dtype='object')

## 🧼 Data Cleaning

The goal of data cleaning is to transform the raw dataset into a structured and reliable format suitable for analysis. Key steps included:
Handle missing values, Drop columns with excessive missing data, Impute missing values using appropriate statistical or categorical strategies, Data Filtering, Remove irrelevant or incomplete records (e.g. non-aircraft incidents or missing key variables), Categorical Grouping, Group low-frequency entries under broader labels (e.g. "Other"), Data Type Correction, Convert date columns to proper datetime format, Ensure numerical columns were correctly typed for analysis, and Outlier Treatment.
These steps ensure we have a clean, consistent, and relevant dataset to move forward with meaningful visualizations and business insights.


In [11]:
#To check missing values in each column
data.isna().sum()

event_id                      0
investigation_type            0
accident_number               0
event_date                    0
location                     52
country                     226
latitude                  54507
longitude                 54516
airport_code              38640
airport_name              36099
injury_severity            1000
aircraft_damage            3194
aircraft_category         56602
registration_number        1317
make                         63
model                        92
amateur_built               102
number_of_engines          6084
engine_type                7077
far_description           56866
schedule                  76307
purpose_of_flight          6192
air_carrier               72241
total_fatal_injuries      11401
total_serious_injuries    12510
total_minor_injuries      11933
total_uninjured            5912
weather_condition          4492
broad_phase_of_flight     27165
report_status              6381
publication_date          13771
dtype: i

### Dropping columns
From data understanding above, we learned that most of our columns were missing values. We will drop all columns missing more than 15%. It may seem very extreme but we chose that percentage since also all of the other columns we don't need are in this percentage. It is important to note that analysis with more than 40% missingness may lead to biasness
 - Example: 
 1. Schedule and air carrier have more than 80% missing values, fillna with mode will bias analysis
 - We may keep far description and aircraft category by filling in unknown 

In [12]:
#identify what to retain
retain= ['far_description','aircraft_category']
#Select the threshold for missing values
threshold= 0.15 
#find columns with 15% or more missing values
dropped_all= data.columns[data.isna().mean()>=threshold]
#exempt the columns to retain
dropped=[col for col in dropped_all if col not in retain]
#drop the identified columns
data= data.drop(columns=dropped)
data.shape

(88889, 23)

### Far description and aircraft category

We will fill the missing values with unknown since most of the values are missing. Filling any other thing will affect the analysis

In [13]:
#fill retain with unknown
[data[col].fillna('unknown', inplace=True)for col in retain]

[None, None]

#### event_id
#### Observation 
We will keep the event_id 
- It has no missing values
- The difference in total and unique events is because some event_ids are similar
- The other values in the rows are not all similar

In [14]:
#we check for duplicates in event_id
missing_events= data['event_id'].isna().sum()
unique_events= data['event_id'].nunique()
total_events = len(data['event_id'])
print(f'There are {unique_events} unique events out of {total_events} total events')
print(f'There are {missing_events} missing events')

There are 87951 unique events out of 88889 total events
There are 0 missing events


#### Investigation type

We will keep this column
- It has 2 values (Accident, Incident)
- It has 0 missing values


In [15]:
#check for missing values
print(data['investigation_type'].isna().sum())
print(data['investigation_type'].nunique())



0
2


#### Accident Number
##### Observation
The column has no missing values
The column has 88863 unique values
We will keep it


In [16]:
data['accident_number'].isna().sum()
data['accident_number'].nunique()

88863

#### Event Date
##### Observation
We will normalize the dates so they are all uniform
- We set the date format to year-month-day
- The column has no missing values

In [17]:
pd.to_datetime(data['event_date'])#y-m-day
data['event_date'].isna().sum()

0

#### Location & Country
- We had 52 missing values that we filled with unknown
- We will merge Us state codes with the location
- We will also fill missing data in country with unknown

In [18]:
#load us_codes
codes = r"C:\Users\USER\Documents\moringa_school\phase_1\phase_1_project\Data\USState_Codes.csv"
us_codes = pd.read_csv(codes, encoding='latin1', low_memory=False)

In [19]:
#standardize location and country using the above dataset
state_dict= dict(zip(us_codes['US_State'].str.lower(),us_codes['Abbreviation']))
data['location']=data['location'].str.lower().map(state_dict).fillna(data['location'])
data['location'].fillna('unknown',inplace=True)
abbrevs= us_codes['Abbreviation'].tolist()
data['country']= data.apply(
     lambda x: 'USA' if x['location'] in abbrevs else x['country'], axis=1
                            )
data['country'].fillna('unknown', inplace=True)
data['location']

0        MOOSE CREEK, ID
1         BRIDGEPORT, CA
2          Saltville, VA
3             EUREKA, CA
4             Canton, OH
              ...       
88884      Annapolis, MD
88885        Hampton, NH
88886         Payson, AZ
88887         Morgan, UT
88888         Athens, GA
Name: location, Length: 88889, dtype: object

### Make, Model & Amateur build

Since the missing values are very small 1% we will fill the mode for categorical values and median for numerical values

In [20]:
#Fill median and mode for <1% missingness
low_missing_cols = data.columns[(data.isna().sum() > 0) & (data.isna().sum() < 0.01 * len(data))]
print(f"Columns with <1% missing values: {list(low_missing_cols)}")
for col in low_missing_cols:
    if data[col].dtype in ['object', 'category']:  # Categorical
        data[col].fillna(data[col].mode()[0], inplace=True)
    else:  # Numerical
        data[col].fillna(data[col].median(), inplace=True)

Columns with <1% missing values: ['make', 'model', 'amateur_built']


#### Injuries

Since total fatal injuries, total serious injuries, total minor injuries,injury severity and total uninjured have moderately high level of missingness, we will fill the missing values with unknown to avoid bias.
Filling 0, will mean that no one had any sort of injury, yet we are not sure

In [21]:
injury_col=['total_fatal_injuries','total_serious_injuries','injury_severity','total_minor_injuries','total_uninjured']
for col in injury_col:
    if col in data.columns:
        data[col].fillna('unknown', inplace=True)

### aircraft damage, registration number, number of engines, engine type, purpose of flight, weather condition, report status

We will also fill unknown for the missin values in the above columns.
For columns  like registration number we cannot just imput any number like a median or median since it is a unique value.

In [22]:
moderate_missing_cols = data.columns[(data.isna().sum() > 1000) & (data.isna().sum() < 0.3 * len(data))]
moderate_missing_cols = [col for col in moderate_missing_cols if col not in retain]
print(f"Columns with >1000 and <30% missing values: {list(moderate_missing_cols)}")
[data[col].fillna('unknown', inplace=True) for col in moderate_missing_cols]


Columns with >1000 and <30% missing values: ['aircraft_damage', 'registration_number', 'number_of_engines', 'engine_type', 'purpose_of_flight', 'weather_condition', 'report_status']


[None, None, None, None, None, None, None]

In [23]:
Verify dataset is clean
data.isna().sum()

SyntaxError: invalid syntax (<ipython-input-23-6d002b8f799f>, line 1)